packages/malloy/src/model/malloy_types.ts

```diff
@@ -884,102 +968,143 @@ export interface TurtleDef extends NamedObject, Pipeline {
+interface SourceDefBase extends StructDefBase, Filtered, ResultStructMetadata {
+  arguments?: Record<string, Argument>;
+  parameters?: Record<string, Parameter>;
+  queryTimezone?: string;
+  connection: string;
+  primaryKey?: PrimaryKeyRef;
 }
+/** which field is the primary key in this struct */
+export type PrimaryKeyRef = string;
 
-export function isJoinOn(sr: StructRelationship): sr is JoinOn {
-  return ['one', 'many', 'cross'].includes(sr.type);
+export interface TableSourceDef extends SourceDefBase {
+  type: 'table';
+  tablePath: string;
 }
-/** types of joins. */
-export type StructRelationship =
-  | {type: 'basetable'; connectionName: string}
-  | JoinOn
-  | {type: 'inline'}
-  | {type: 'nested'; fieldName: string; isArray: boolean};
 
+/*
+ * Malloy has a kind of "strings" which is a list of segments. Each segment
+ * is either a string, or a query, which is meant to be replaced
+ * by the text of the query when the query is compiled to SQL.
+ *
+ * The data types for this are:
+ *  SQLPhrase -- A phrase, used to make a sentence
+ *  SQLSentence -- Used to request a schema from the connection
+ *  SQLSelectStruct -- Returned from a query, contains the scehma
+ */
 export interface SQLStringSegment {
   sql: string;
 }
-export type SQLPhrase = Query | SQLStringSegment;
-export function isSQLFragment(f: SQLPhrase): f is SQLStringSegment {
-  return (f as SQLStringSegment).sql !== undefined;
+export type SQLPhraseSegment = Query | SQLStringSegment;
+export function isSegmentSQL(f: SQLPhraseSegment): f is SQLStringSegment {
+  return 'sql' in f;
 }
-/**
- * A source reference to an SQL block. The compiler uses these to request
- * an SQLBlock with it's schema and structdef defined. Use the factory
- * makeSQLBlock to construct these.
- */
-export interface SQLBlockSource {
+
+export interface SQLSentence {
   name: string;
-  connection?: string;
-  select: SQLPhrase[];
+  connection: string;
+  select: SQLPhraseSegment[];
 }
 
-export interface SQLBlock extends NamedObject {
-  type: 'sqlBlock';
-  connection?: string;
+export interface SQLSourceDef extends SourceDefBase {
+  type: 'sql_select';
   selectStr: string;
 }
 
-interface SubquerySource {
-  type: 'sql';
-  method: 'subquery';
-  sqlBlock: SQLBlock;
+export interface QuerySourceDef extends SourceDefBase {
+  type: 'query_source';
+  query: Query;
 }
 
-/** where does the struct come from? */
-export type StructSource =
-  | {type: 'table'; tablePath: string}
-  | {type: 'nested'}
-  | {type: 'inline'}
-  | {type: 'query'; query: Query}
-  | {type: 'sql'; method: 'nested' | 'lastStage'}
-  | {type: 'query_result'}
-  | SubquerySource;
+export interface QueryResultDef extends SourceDefBase {
+  type: 'query_result';
+}
 
-// Inline and nested tables, cannot have a StructRelationship
-//  the relationshipo is implied
+// Describes the schema which flows between pipe elements
+export interface NestSourceDef extends SourceDefBase {
+  type: 'pipeline_struct';
```

Comment:
`nest_source`?

Reply:
yeah i was thinking that too .. again i didn't know what these were when i started